clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -517,6 +519,50 @@ public static int murmur2(final byte[] data) {
         return h;
     }
 
+    /**
+     * Generates 32 bit murmur2 hash from ByteBuffer
+     * @param data ByteBuffer to hash
+     * @return 32 bit hash of the given ByteBuffer
+     */
+    @SuppressWarnings("fallthrough")
+    public static int murmur2(ByteBuffer data) {
```

Comment:
I added MurMurHashBenchmark: 
```
Benchmark                              (bytes)  (direct)  (readonly)   Mode  Cnt      Score     Error   Units
MurMurHashBenchmark.byteArrayMurmur2       128     false       false  thrpt   10  23899.296 ± 286.754  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       128     false        true  thrpt   10  23967.434 ± 175.354  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       128      true       false  thrpt   10  23809.852 ± 285.063  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       128      true        true  thrpt   10  23897.052 ± 270.672  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       256     false       false  thrpt   10  12312.813 ±  88.589  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       256     false        true  thrpt   10  12321.595 ±  90.111  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       256      true       false  thrpt   10  12305.468 ±  79.264  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       256      true        true  thrpt   10  12324.379 ±  83.018  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      128     false       false  thrpt   10  20663.903 ± 149.945  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      128     false        true  thrpt   10  20586.696 ± 283.857  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      128      true       false  thrpt   10  26596.222 ± 638.404  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      128      true        true  thrpt   10  26719.416 ± 405.441  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      256     false       false  thrpt   10  11270.697 ±  89.376  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      256     false        true  thrpt   10  11015.074 ± 103.103  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      256      true       false  thrpt   10  14733.461 ± 116.067  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      256      true        true  thrpt   10  14733.619 ± 250.710  ops/ms

```

Reply:
> Seems like we have a 13% regression for non-direct buffer cases and similar improvement in direct buffer cases. The additional computation we are performing for non-direct buffer is `data.arrayOffset() + data.position()` which is done once (hopefully, compiler is intelligent enough to that). Could we perhaps take it out of the loop just to be double sure.
> 
> Any other thoughts on why do we have a perf difference between direct vs non-direct buffer here? I expect a performance difference between them when GC is involved but in this case we are not creating more objects (except when calling slice()).

I believe that `data.arrayOffset() + data.position()` is not the key issue, but rather `ByteBuffer#getInt(int)`:
* In aarch64/x86_64/x86/ppc64le/ppc64/i386 architectures, `DirectByteBuffer` uses `Unsafe#getInt(int)` to retrieve 4 bytes in one operation.
* No matter what architectures, `HeapByteBuffer` uses `ByteBuffer#_get(int)` to  retrieve 4 bytes in four operations.

Reply:
That's fair. Prior to this change byte[] would also have been stored contiguously hence reading 4 bytes would have been faster whereas HeapByteBuffer may perform random seeks to read the entire int. 

It's fine since I don't see how we can potentially avoid this but note that this may offset gains that we are getting by removing the data copy in this PR.

Reply:
I believe that the performance impact of this part will only slightly diminish the optimizations brought by this PR, rather than completely offsetting them. This is because Utils#murmur2 is only applied to calculate the partition to which a key belongs, and the key is often much smaller than the value.

Reply:
@divijvaidya 
I optimized `Utils#murmur2(ByteBuffer)` through this [commit](https://github.com/apache/kafka/pull/12685/commits/3dec26dadf60c43e53ff8b13ef83206aabf82637), and here are the results:
```
Benchmark                              (bytes)  (direct)  (littleEndian)  (seed)   Mode  Cnt      Score     Error   Units
MurMurHashBenchmark.byteArrayMurmur2       128     false           false      42  thrpt   10  23952.036 ± 234.049  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       128     false            true      42  thrpt   10  24006.164 ± 153.647  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       128      true           false      42  thrpt   10  23997.223 ± 155.545  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       128      true            true      42  thrpt   10  24014.765 ± 169.769  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       256     false           false      42  thrpt   10  12314.915 ±  71.558  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       256     false            true      42  thrpt   10  12312.495 ±  87.004  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       256      true           false      42  thrpt   10  12299.405 ±  85.441  ops/ms
MurMurHashBenchmark.byteArrayMurmur2       256      true            true      42  thrpt   10  12289.485 ± 107.814  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      128     false           false      42  thrpt   10  23436.309 ± 169.624  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      128     false            true      42  thrpt   10  23436.878 ±  93.588  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      128      true           false      42  thrpt   10  25995.246 ± 582.841  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      128      true            true      42  thrpt   10  28573.727 ± 155.530  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      256     false           false      42  thrpt   10  12210.036 ±  69.601  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      256     false            true      42  thrpt   10  12184.773 ±  98.403  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      256      true           false      42  thrpt   10  14500.743 ± 196.826  ops/ms
MurMurHashBenchmark.byteBufferMurmur2      256      true            true      42  thrpt   10  15421.734 ± 113.717  ops/ms
```

After the optimization, the throughput of `Utils#murmur2(ByteBuffer)` in `HeapByteBuffer` **decreased** by only **2%** compared to `Utils#murmur2(byte[])`, while in `DirectByteBuffer`, the throughput **increased** by **8%**.